src/adapter/http/src/data/metadata_handler.rs

```diff
@@ -0,0 +1,230 @@
+pub struct ResponseBody {
+    pub output: Output,
+}
+
+#[serde_with::serde_as]
+#[serde_with::skip_serializing_none]
+#[derive(Debug, serde::Serialize)]
+#[serde(rename_all = "camelCase", deny_unknown_fields)]
+pub struct Output {
+    #[serde_as(as = "Option<odf::serde::yaml::SetAttachmentsDef>")]
+    pub attachments: Option<odf::SetAttachments>,
+
+    #[serde_as(as = "Option<odf::serde::yaml::SetInfoDef>")]
+    pub info: Option<odf::SetInfo>,
+
+    #[serde_as(as = "Option<odf::serde::yaml::SetLicenseDef>")]
+    pub license: Option<odf::SetLicense>,
+
+    pub refs: Option<Vec<Ref>>,
+
+    pub schema: Option<String>,
+    pub schema_format: Option<query_types::SchemaFormat>,
+
+    #[serde_as(as = "Option<odf::serde::yaml::SeedDef>")]
+    pub seed: Option<odf::Seed>,
+
+    #[serde_as(as = "Option<odf::serde::yaml::DatasetVocabularyDef>")]
+    pub vocab: Option<odf::DatasetVocabulary>,
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+#[derive(Debug, serde::Serialize)]
+#[serde(rename_all = "camelCase", deny_unknown_fields)]
+pub struct Ref {
+    pub name: String,
+    pub block_hash: odf::Multihash,
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+#[transactional_handler]
+pub async fn dataset_metadata_handler(
+    Extension(catalog): Extension<Catalog>,
+    Extension(dataset_ref): Extension<odf::DatasetRef>,
+    Query(params): Query<RequestParams>,
+) -> Result<Json<ResponseBody>, ApiError> {
+    use kamu_core::{metadata_chain_visitors as vis, MetadataChainExt as _};
+
+    let dataset_repo = catalog.get_one::<dyn DatasetRepository>().unwrap();
+    let dataset = dataset_repo
+        .find_dataset_by_ref(&dataset_ref)
+        .await
+        .api_err()?;
+
+    let mut attachments_visitor = params
+        .include
+        .contains(&Include::Attachments)
+        .then(vis::SearchSetAttachmentsVisitor::new);
+    let mut info_visitor = params
+        .include
+        .contains(&Include::Info)
+        .then(vis::SearchSetInfoVisitor::new);
+    let mut license_visitor = params
+        .include
+        .contains(&Include::License)
+        .then(vis::SearchSetLicenseVisitor::new);
+    let mut schema_visitor = params
+        .include
+        .contains(&Include::Schema)
+        .then(vis::SearchSetDataSchemaVisitor::new);
+    let mut seed_visitor = params
+        .include
+        .contains(&Include::Seed)
+        .then(vis::SearchSeedVisitor::new);
+    let mut vocab_visitor = params
+        .include
+        .contains(&Include::Vocab)
+        .then(vis::SearchSetVocabVisitor::new);
+
+    let mut visitors: [&mut dyn MetadataChainVisitor<Error = vis::Infallible>; 6] = [
+        &mut attachments_visitor,
+        &mut info_visitor,
+        &mut license_visitor,
+        &mut schema_visitor,
+        &mut seed_visitor,
+        &mut vocab_visitor,
+    ];
```

Comment:
FYI @s373r I'm super happy about this trick:
- Notice that all response elements are optional - so if user did not request `seed` or `schema` we ideally don't want to search for this event
- So I made all visitors above optional
- I then implemented `MetadataChainVisitor` for `Option<T> where T: MetadataChainVisitor`
- If the `Option` is `None` - the visitor will immediately `Stop`
- If `Some` - it delegates the calls to inner visitor

This is about as concise as it gets :)

Reply:
@sergiimk , :100: 

Awesome idea with the optional visitors -- looking at the changes in the other PR noticed this, very cool!